Feat aging indicator

smattingly opened the pull request:
This implements [Aging Indicator and filter/report](https://airtable.com/tblk2ToSNYzNVza5f/viwrmqbUpwcUGA2OM/recLKBmAEWROG52HP)

![image](https://user-images.githubusercontent.com/21227171/129112836-2e5a4114-644c-47d5-8559-84f931fee930.png)

In all views that list grants, the `danger` modifier (red background) is applied to close dates on or before the current date. The `warning` modifier (yellow background) is applied to close dates less that X days in the future, where X is coded to a default value of 21 but *should* override that with an environment variable's value if present.

*Help needed* 
1. ~~I can't get the environment variable to work. I'm new to Vue, and I think the code is okay but that I'm not setting up the environment correctly?~~ Resolved; thanks, Michael.
2. I have added the "Aging" view shown above but that's probably not the right place, and maybe not all the content. I need some help interpreting:
> Additionally, if a grants administrator had a view or report that showed all grants across agencies closing in the same period as above, it would be very helpful for following up before it's too late.

hcdeng commented on packages/client/src/components/GrantsTable.vue:
```diff
@@ -138,6 +138,18 @@ export default {
+        const agingThreshold = 1000 * 60 * 60 * 24 * (process.env.VUE_APP_AGING_THRESHOLD_DAYS || 21);
+        const result = {};
+        const diff = new Date(grant.close_date) - new Date();
+        if (diff <= 0) {
```
Probably a question for the partners, but I wonder if it would be more helpful to gray out grants that are already closed, and highlight in red at a different threshold, like 7 days or something.

smattingly replied:
Good point. I'm not certain if closed grants will actually show in the application anyway. There are closed grants at grants.gov, and the cron job does not seem to load those. But I'm not certain if there is a mechanism to purge loaded grants after they have closed. @Rapol ?

Rapol commented:
The aging indicator is shown no matter where the table is displayed, correct? If so what is the different between My Grants interested and aging tabs? 

Other than that looks really good. The aging indicator at the agency level is a good idea

smattingly commented:
Per partner feedback, I have stored close date thresholds for each partner (currently, each record in the `agencies` table.). Default values are 30 days for warning and 15 days for danger. Admin users can modify these values for the agencies that they can access. A user's agency (as indicated in their `users` table record) determine what threshold values are used in their display. 

![image](https://user-images.githubusercontent.com/21227171/129994176-3d372050-8b1c-4cdd-b213-89b492433239.png)

I would appreciate feedback on UI clarity and if there are better Vue ways to do this.

Also, I removed the "aging" view that I had added since that didn't seem to be what the partners had in mind.

Rapol commented:
Dialog looks perfect and Vue code is very clean.

I would prefer a put agencies/:id instead of put agencies/:id/thresholds just in case in the future we add more fields to the dialog but we should leave it like we have it now. No need to change it until we have new fields to add.

smattingly commented:
> I would prefer a put agencies/:id instead of put agencies/:id/thresholds just in case in the future we add more fields to the dialog but we should leave it like we have it now. No need to change it until we have new fields to add.

I agree, so I will go ahead and change that. I usually try to stick to [this](http://microformats.org/wiki/rest/urls#Routing) as closely as I can, but I saw 
```fetchApi.put(`/api/eligibility-codes/${code}/enable/${enabled}`);``` for example, and wrote it more "specifically".

hcdeng commented:
This looks awesome! Maybe you already addressed with the partners, but I'm thinking that anywhere you say "Warning" or "Danger" you could replace or include the actual color to be clear to users. So, "Close Date Warning (Yellow) Threshold" for example.

smattingly commented:
@Rapol I just changed the URL as you described. I did not change any of the function names, etc.

smattingly commented:
> This looks awesome! Maybe you already addressed with the partners, but one UI suggestion from me: anywhere you say "Warning" or "Danger" you could replace or include the actual color to be clearer to users. So, "Close Date Warning (Yellow) Threshold" for example.

@hcdeng I see your point, and I made a change to apply the styles to the description text:
![image](https://user-images.githubusercontent.com/21227171/130135915-744728f5-bf0b-4b68-8022-0631eb526a0a.png)

Unfortunately, it's kind of small.

I'm no expert, but I feel that using color names might not be good for accessibility. The same argument applies to using the colors themselves to reflect status, but strictly speaking I'm not using the colors-- I'm applying the `text-warning` or `text-danger` classes, which could be styled in various ways and (I hope) could be useful to accessibility tools.

hcdeng commented:
@smattingly Great point - makes sense to me!

smattingly commented:
@Rapol I just realized that the Agencies view shows the user *all* agencies; I thought it showed only the user's own agency plus its ancestors and/or descendants via the `parent` association. 
So, should an admin user be able to change the thresholds for all/any agencies? (If not, I think this is related or even identical to some of what @Michael-Stanford has described for centralization.)

Rapol commented:
I think we can tackle that later. User and agencies right now are both global, meaning that an admin can modify them.

Rapol commented:
There are some merge conflicts with yarn.lock. Lets accept the changes from develop

smattingly commented:
> There are some merge conflicts with yarn.lock. Lets accept the changes from develop

As mentioned in Slack, I can't cleanly merge develop due to lint errors.
